src/vs/workbench/contrib/timeline/common/timelineService.ts

```diff
@@ -0,0 +1,207 @@
+		// 				description: uri.toString(true),
+		// 				date: Date.now()
+		// 			},
+		// 			{
+		// 				id: '2',
+		// 				label: 'Undo Timeline2',
+		// 				description: uri.toString(true),
+		// 				date: Date.now() - 100
+		// 			}
+		// 		];
+		// 	},
+		// 	dispose() { }
+		// });
+	}
+
+	async getTimeline(uri: URI, token: CancellationToken, sources?: Set<string>) {
+		this.logService.trace(`TimelineService#getTimeline(${uri.toString(true)})`);
+
+		const requests = new Map<string, Promise<TimelineItem[] | CancellationErrorWithId<string>>>();
+
+		for (const provider of this._providers.values()) {
+			if (sources && !sources.has(provider.source)) {
+				continue;
+			}
+
+			requests.set(provider.source, provider.provideTimeline(uri, token));
+		}
+
+		// TODO[ECA]: What should the timeout be for waiting for individual providers?
+		const timelines = await raceAll(requests /*, 5000*/);
+
+		const timeline = [];
+		for (const [source, items] of timelines) {
+			if (items instanceof CancellationError) {
+				this.logService.trace(`TimelineService#getTimeline(${uri.toString(true)}) source=${source} cancelled`);
+				continue;
+			}
+
+			if (items.length === 0) {
+				continue;
+			}
+
+			timeline.push(...items.map(item => ({ ...item, source: source })));
+		}
+
+		timeline.sort((a, b) => b.timestamp - a.timestamp);
+		return timeline;
+	}
+
+	registerTimelineProvider(provider: TimelineProvider): IDisposable {
+		this.logService.trace(`TimelineService#registerTimelineProvider: source=${provider.source}`);
+
+		const source = provider.source;
+
+		const existing = this._providers.get(source);
+		if (existing && !existing.replaceable) {
+			throw new Error(`Timeline Provider ${source} already exists.`);
+		}
+
+		this._providers.set(source, provider);
+		if (provider.onDidChange) {
+			this._providerSubscriptions.set(source, provider.onDidChange(uri => this.onProviderTimelineChanged(provider.source, uri)));
+		}
+		this._onDidChangeProviders.fire();
+
+		return {
+			dispose: () => {
+				this._providers.delete(source);
+				this._onDidChangeProviders.fire();
+			}
+		};
+	}
+
+	unregisterTimelineProvider(source: string): void {
+		this.logService.trace(`TimelineService#unregisterTimelineProvider: source=${source}`);
+
+		if (!this._providers.has(source)) {
+			return;
+		}
+
+		this._providers.delete(source);
+		this._providerSubscriptions.delete(source);
+		this._onDidChangeProviders.fire();
+	}
+
+	private onProviderTimelineChanged(source: string, uri: URI | undefined) {
+		// console.log(`TimelineService.onProviderTimelineChanged: source=${source} uri=${uri?.toString(true)}`);
+
+		this._onDidChangeTimeline.fire(uri);
+	}
+}
+
+function* map<T, TMapped>(source: Iterable<T> | IterableIterator<T>, mapper: (item: T) => TMapped): Iterable<TMapped> {
+	for (const item of source) {
+		yield mapper(item);
+	}
+}
+
+class CancellationError<TPromise = any> extends Error {
+	constructor(public readonly promise: TPromise, message: string) {
```

Comment:
We usually use `errors.ts#cancled`

Reply:
Yeah, and we usually treat cancellation as a rejection, not a resolution. The current code is a bit hard to understand, since it seems that promises are resolving with errors instead of rejecting with them. I see how here there's a use case for N promises to resolve and some might reject but not block the overall race... but instead of resolving using a generic `CancellationErrorWithId` why not resolving with a specific `NoResults` result type instead?

Reply:
Basically I am racing each promise with a timeout so that any one provider can't take too long to return results (although currently I have that turned off as there is no timeout specified). I'm not 100% sure if we do want to do this (imo we probably should have some limit), and if we do, I probably need to think more about what to do with the requests that take too long. Do we cancel them? Do we render everything that is ready and then continue to wait longer for them to finish and then update the UI with those additional results?

I lifted this code from GitLens (so it has some extra needs baked into it), e.g. the `CancellationErrorWithId` contains both a correlation id (to map up with requests) and the promise that we stopped waiting for (its still running).

Reply:
For format on save and code actions on save I went away from timeouts because machines vary too much and also expectation vary too much. Instead, it now shows progress after a short time and that allows to cancel the operation. I came to believe that's more user friendly then timeouts and settings for those timeouts